llvm/include/llvm/ADT/StringMap.h

```diff
@@ -257,6 +257,15 @@ class LLVM_ALLOCATORHOLDER_EMPTYBASE StringMap
     return ValueTy();
   }
 
+  /// lookup - Return the entry for the specified key, or a default
+  /// provided value if no such entry exists.
+  const ValueTy &lookup(StringRef Key, const ValueTy& Value) const {
```

Comment:
This seems like quite a footgun to me: the lifetime of the default value won't be extended beyond the lookup expression, whereas the looked up object will be kept alive by the map.

Consider this code:
```c++
StringMap<SmallVector<int>> foo;
const auto &ints = foo.lookup("X", SmallVector<int>{});
```

I think `ints` will be a dangling reference.

Reply:
I agree that you could use this in a wrong way, but I'm not sure it's not true for many other use cases we're already familiar with.

In the future, we could probably make this more safe if we use the `lifetimebound` attribute.

For examples of APIs that rely on a parameter outliving the returned value, a quick in ADT search shows there are examples (partial list, and I only searched for const reference cases):
* https://github.com/llvm/llvm-project/blob/7844257fc2afe490ae4b923a770d20dabed5c3c6/llvm/include/llvm/ADT/APInt.h#L2217
* https://github.com/llvm/llvm-project/blob/7844257fc2afe490ae4b923a770d20dabed5c3c6/llvm/include/llvm/ADT/ConcurrentHashtable.h#L89

Reply:
For example, `std::optional::default_or` returns by value, which seems more reasonable to me: https://en.cppreference.com/w/cpp/utility/optional/value_or

Reply:
I think one safer alternative would be to return the looked up value through a pointer

Reply:
I was aiming to return by reference following the discussion in https://github.com/llvm/llvm-project/pull/112517#discussion_r1832610617.
When the value is costly to copy, even if the default value is cheap, returning by value would be costly.
I assume there would be a lot of use cases for this and it could simplify the code.
I don't feel strongly about the name.
Returning a pointer would work, though it seems to be somewhat less elegant. Do you mean that also the default value would be a pointer?

Reply:
> I assume there would be a lot of use cases for this and it could simplify the code.

Is this more of a guess or have you identified multiple places where it'd help across the project?

> Returning a pointer would work, though it seems to be somewhat less elegant. Do you mean that also the default value would be a pointer?

I was thinking of something like `auto* val = x.lookup_or(Key, ptr)` or `x.lookup(Key, ptrPtr);` (where `ptrPtr` is written to if the value is present, otherwise it keeps pointing to whatever it used to). This is not very elegant but it's very explicit around lifetimes, but I think it should be a better option for very specialized use cases that warrant it.

Reply:
IME such `lookup` APIs are usually implemented by having a static default constructed object T and returning a reference to it when the find fails.

This surely has some downsides like never running the destructor until program exit for such objects and eating more memory as people instantiate more stringmaps :/ So I am not sure if it's worth the simplicity of code.

Reply:
Thanks everyone!
I agree this API is not ideal and might require a stronger use case to introduce, so closing this PR.